### ISSUE TITLE
Support Unicode labels

### ISSUE THREAD

**Issue by sellout** [dhall-lang/dhall-lang#487]
I know that there is concern of obfuscation from allowing Unicode, but I find it really useful and can imagine that plenty of people would at least prefer to use languages other than English. So, barring a switch from `ALPHA` to “Letter” and other similar changes, I have some suggestions.

1. expand `quoted-label-char` similarly to `quoted-path-character`,
2. expand `ALPHA` with some “vetted” ranges of non- obfuscatory characters (I’d vote for some linguistic ranges as well as mathematical alphabets), and
3. add some ranges (like super- & subscripts) to `simple-label-next-char`.

**Comment by Gabriel439** [comment]
@sellout: Yeah, I'm in favor of this.  In particular, now that we support α-normalization we can easily foil obfuscation attempts when auditing suspicious code.

There are two main things things we need to decide on:

* what characters to permit
* how to determine equivalence of Unicode identifiers

The simplest possible solution is:

* Permit all characters between `%x20` and `%x10FFFF`
* Treat identifiers as equivalent if they have equivalent code points

This is what I propose for now to simplify the implementation burden.  This is also consistent with how we already handled quoted paths.

However, another solution which I'll mention for completeness is to do something more sophisticated, like in this document:

http://www.unicode.org/reports/tr31/tr31-31.html

That includes things like:

* ruling out weird characters like zero-width-joiners or bidirectional ordering control
* detecting equivalent Unicode strings
* protecting against visual confusion (when not α-normalized)

I'm also in favor of expanding the grammar for unquoted labels to permit subscripts and superscripts.  I've been wanting this for myself for a while.

However, I think we should keep `ALPHA` the same, at least for now, because it will increase our maintenance burden while people debate which scripts to include.  We should wait to do that when there are fewer other changes in flight.

**Comment by singpolyma** [comment]
‎> * ruling out weird characters like zero-width-joiners or bidirectional ordering control

I think at least this should be included -- whitespace/control points like ZWJ should not be part of the identifier space

> * detecting equivalent Unicode strings

I think using canonicalized Unicode for identifiers is essential if allowing them -- otherwise identifiers meant to be the same and visually identical ‎to the author could be not resolved together and confuse the heck out of the author because of internal tech details like "codepoints" which would not be human friendly.

**Comment by Gabriel439** [comment]
@singpolyma: Yeah, so on more reflection I think ruling out weird characters shouldn't be too hard as long as we can provide the ABNF for the correct character ranges.  I'm still looking for a reference that spells out the character ranges for each Unicode character class because I can only find explicit lists of characters at the moment.

For canonicalizing Unicode identifiers, I think that's fine if it's not hard for other languages to implement.  I'm not that familiar with Unicode implementations in other languages, so I didn't want to presume how difficult it was.

**Comment by Gabriel439** [comment]
Does anybody know a resource for which character ranges correspond to which Unicode general categories?  The part I get stuck on when I try to do this the right way is that I can't figure how to turn the general categories recommended for identifiers in http://www.unicode.org/reports/tr31/tr31-31.html to ABNF rules

**Comment by philandstuff** [comment]
A bit of googling found [this page](
http://unicode.org/cldr/utility/list-unicodeset.jsp?abb=on&esc=on&a=%5B%3AID_Start%3A%5D%5D)
which *appears* to list all the ID_Start characters, abbreviating
contiguous ranges

**Comment by sjakobi** [comment]
> Does anybody know a resource for which character ranges correspond to which Unicode general categories?

I've made [a little module](https://github.com/sjakobi/unicode-character-ranges/blob/master/src/Lib.hs) that derives this mapping from `Data.Char.generalCategory`. If you use it with GHC-8.8, the results should be up-to-date with Unicode v12.

**Comment by Gabriel439** [comment]
Looking into this further, it looks like the approach I'll probably take is to implement the following sections from [UAX #31](http://www.unicode.org/reports/tr31/tr31-31.html), which I think strike a good balance between being easy to standardize/implement and behaving well:

* [R1. Default identifiers](http://www.unicode.org/reports/tr31/tr31-31.html#R1)

  This specifies a sensible default set of characters permitted in identifiers (which we can tweak as necessary)

  Specifically, we'd use `XID_Start XID_Continue*` as the grammar rule for unquoted identifiers

* [R3. Pattern_White_Space and Pattern_Syntax Characters](http://www.unicode.org/reports/tr31/tr31-31.html#R3)

  This specifies a sensible default set of characters reserved for whitespace and operators (which we can tweak as necessary)

* [R4. Equivalent Normalized Identifiers](http://www.unicode.org/reports/tr31/tr31-31.html#R4)

  This specifies that identifiers don't have to be in normal form in the code, but they are converted to a normal form of our choice when comparing for equality.  The above standard recommends using normal form C (NFC) if identifiers are case-sensitive.

For more details on normal forms, see [UAX #15](https://unicode.org/reports/tr15/)

However, that still leaves open one issue, which is that the above standard provides no guidance on what set of characters should be permitted for quoted identifiers.  Specifically, `XID_CONTINUE+` would probably not work since it does not permit whitespace and punctuation.  We currently permit punctuation within quoted identifiers and probably want to permit whitespace within quoted identifiers, too.

I'm guessing we probably want to permit `(XID_CONTINUE | Pattern_White_Space | Pattern_Syntax)+` within quoted identifiers.

If people don't object to that I'll begin drafting a proposal based on that.

**Comment by sellout** [comment]
Yeah, I like this – I couldn’t easily figure out if I get to use super-/subscripts, but at least I get all my mathematical letters – script, bold, double-struck, etc. So I’m happy. And I like the `Start = XID_Start` approach, where you can then incrementally add onto that default set.
